### PR TITLE
[Enhancement]: Exporting some models to OpenVINO

### PR DESCRIPTION
**Pull request by SemyonBevzuk**
## Motivation

It has been found that some of the models are simple enough to export to OpenVINO.
The mmdetection already has experimental support for exporting to TensorRT, so this PR adds functionality for exporting some models to OpenVINO.
Supported models:
- FCOS
- FSAF
- RetinaNet
- SSD
- YOLOv3
- Faster RCNN
- Cascade RCNN
- Mask RCNN
- Cascade Mask RCNN

## Modification

- `tools/deployment/`
   `tools/deployment/pytorch2onnx.py`: Argument parsing moved to function. Added a main function for easy code reuse in other scripts.
   `tools/deployment/pytorch2openvino.py`: Added a script for export to OpenVINO.
- `mmdet/core/export/openvino_wrapper/`: This directory contains the scripts needed for export to OpenVINO.
  `mmdet/core/export/openvino_wrapper/openvino_helper.py`: Contains the OpenvinoExportHelper class that applies symbolic functions.
  `mmdet/core/export/openvino_wrapper/symbolic.py`: Additional symbolics and tools for working with them.
- `mmdet/core/export`
   `mmdet/core/export/pytorch2onnx.py`: Added support for device type (CPU or CUDA) in `preprocess_example_input` for `one_img`
   `mmdet/core/export/onnx_helper.py`: The `device` parameter has been added to `torch.arange`.
- `mmdet/models/dense_heads/`
   `mmdet/models/dense_heads/fcos_head.py`: The `device` parameter has been added to `torch.arange`.
   `mmdet/models/dense_heads/rpn_head.py`: The `device` parameter has been added to `torch.arange`.
   `mmdet/models/dense_heads/yolo_head.py`: The `device` parameter has been added to `torch.arange`.
- `configs/_base_/models/mask_rcnn_r50_fpn.py`: The `rescale_mask_to_input_shape_in_onnx` parameter has been added to skip postprocessing.
- `mmdet/models/roi_heads/`
   `mmdet/models/roi_heads/mask_heads/fcn_mask_head.py`: Added the ability to skip postprocessing.
   `mmdet/models/roi_heads/cascade_roi_head.py`: Fixed problems with shape when skipping postprocessing.
   `mmdet/models/roi_heads/standard_roi_head.py`: Fixed problems with shape when skipping postprocessing.
- `docs/tutorials/pytorch2openvino.md`: Documentation has been added.

## Use cases (Optional)
Before exporting a model, you need to install [OpenVINO](https://docs.openvinotoolkit.org/latest/installation_guides.html).
After that you can run the script `tools/deployment/pytorch2openvino.py` with arguments, that are similar to `tools/deployment/pytorch2onnx.py`.
This script will apply additional symbolics and patches, export the model to ONNX and call the Model Optimizer (mo.py) to export the model to OpenVINO.

Example:
```bash
python /home/sbevzyuk/openmmlab_dir/mmdetection/tools/deployment/pytorch2openvino.py /tmp/openmmlab/ssd/ssd300_coco/config.py /tmp/openmmlab/snapshots/ssd300_coco_20210803_015428-d231a06e.pth --output-file /tmp/openmmlab/ssd/ssd300_coco/config.onnx --opset-version 11 --dynamic-export --not_strip_doc_string
```
Output:
```
         Symbolic function patches:
Patch roi_feature_extractor applied

{Messages from pytorch2onnx.py}

Successfully exported ONNX model: /tmp/openmmlab/ssd/ssd300_coco/config.onnx

Args for mo.py: mo.py --input_model="/tmp/openmmlab/ssd/ssd300_coco/config.onnx" --mean_values="[123.675, 116.28, 103.53]" --scale_values="[1, 1, 1]" --output_dir="/tmp/openmmlab/ssd/ssd300_coco" --output="labels,dets" --input_shape="[1, 3, 300, 300]" --reverse_input_channels --disable_fusing

{Messages from OpenVINO mo.py}

Successfully exported OpenVINO model: /tmp/openmmlab/ssd/ssd300_coco/config.xml, /tmp/openmmlab/ssd/ssd300_coco/config.bin
```

### CODE REVIEW

**Comment by ZwwWayne** [comment]
Hi @SemyonBevzuk ,
Thank you for the PR.OpenVINO is an important feature of mmdet. We will review it and hopefully we could merge it. Could you fix the lint issue first?

File: tools/deployment/pytorch2onnx.py [hunk 1]
```diff
@@ -30,8 +30,13 @@ def pytorch2onnx(model,
...
         'normalize_cfg': normalize_cfg
     }
+
+    if torch.cuda.is_available():
+        model.cuda()
+    device = next(model.parameters()).device
+
     # prepare input
-    one_img, one_meta = preprocess_example_input(input_config)
+    one_img, one_meta = preprocess_example_input(input_config, device)
```
Comment (RunningLeon) [inline tools/deployment/pytorch2onnx.py:39]: `device` argument is missed in  Line 149.

Comment (SemyonBevzuk) [reply tools/deployment/pytorch2onnx.py:39]: Fixed.

File: mmdet/core/export/pytorch2onnx.py [hunk 2]
```diff
@@ -102,11 +102,13 @@ def build_model_from_cfg(config_path, checkpoint_path, cfg_options=None):
     return model
 
 
-def preprocess_example_input(input_config):
+def preprocess_example_input(input_config, device=torch.device('cpu')):
```
Comment (RunningLeon) [inline mmdet/core/export/pytorch2onnx.py:105]: Have you tested exporting to ONNX with `device=cuda`?

**Comment by RangiLyu** [comment]
Should the `openvino_wrapper` folder be moved into `mmdet/core/export`? @RunningLeon Any suggestions?

**Comment by RunningLeon** [comment]
> Should the `openvino_wrapper` folder be moved into `mmdet/core/export`? @RunningLeon Any suggestions?

@RangiLyu Yes, seems better to me.

File: mmdet/core/export/pytorch2onnx.py (hunk 2 above)
Comment (SemyonBevzuk) [reply mmdet/core/export/pytorch2onnx.py:106]: I will check it.

Comment (SemyonBevzuk) [reply mmdet/core/export/pytorch2onnx.py:106]: It looks like everything works correctly with `device=torch.device('cuda:0')`.

**Review by RangiLyu — APPROVED** [review]
LGTM

File: configs/_base_/models/mask_rcnn_r50_fpn.py [hunk 3]
```diff
@@ -117,4 +117,5 @@
             score_thr=0.05,
             nms=dict(type='nms', iou_threshold=0.5),
             max_per_img=100,
-            mask_thr_binary=0.5)))
+            mask_thr_binary=0.5,
+            rescale_mask_to_input_shape=True)))
```
Comment (jshilong) [inline configs/_base_/models/mask_rcnn_r50_fpn.py:121]: Seems we can add a comment here to indicate this argument only works when exporting to ONNX or support it in normal forward to be consistent.

**Review by jshilong — APPROVED** [review]
Can be merged after the last comment is solved

File: configs/_base_/models/mask_rcnn_r50_fpn.py (hunk 3 above)
Comment (SemyonBevzuk) [reply configs/_base_/models/mask_rcnn_r50_fpn.py]: Changed the parameter name from `rescale_mask_to_input_shape` to `rescale_mask_to_input_shape_in_onnx`. I think this should improve readability and should not be misleading when using normal forward.

**Comment by SemyonBevzuk** [comment]
Can you help with failing **build/build cpu (3.7, 1.5.1)**?
Locally tests are green.
When launched in СI, the following problem occurs:
```
/home/runner/work/_temp/221e346b-5901-476a-ade0-932266282857.sh: line 1:  2289 Killed                  coverage run --branch --source mmdet -m pytest tests/
tests/test_models/test_forward.py 
Error: Process completed with exit code 137.
```
I do not quite understand what is the reason for it and how to fix it.

**Comment by RangiLyu** [comment]
> Can you help with failing **build/build cpu (3.7, 1.5.1)**?
> Locally tests are green.
> When launched in СI, the following problem occurs:
> 
> ```
> /home/runner/work/_temp/221e346b-5901-476a-ade0-932266282857.sh: line 1:  2289 Killed                  coverage run --branch --source mmdet -m pytest tests/
> tests/test_models/test_forward.py 
> Error: Process completed with exit code 137.
> ```
> 
> I do not quite understand what is the reason for it and how to fix it.

`Error: Process completed with exit code 137.` means out of memory. MMDetection's github CI always be killed because of the memory issue. And we are working on solving this problem.

**Review by RunningLeon — APPROVED** [review]
LGTM

File: configs/_base_/models/mask_rcnn_r50_fpn.py (hunk 3 above)
Comment (SemyonBevzuk) [reply configs/_base_/models/mask_rcnn_r50_fpn.py]: Kindly ping @jshilong

**Comment by RangiLyu** [comment]
> Can you help with failing **build/build cpu (3.7, 1.5.1)**?
> Locally tests are green.
> When launched in СI, the following problem occurs:
> 
> ```
> /home/runner/work/_temp/221e346b-5901-476a-ade0-932266282857.sh: line 1:  2289 Killed                  coverage run --branch --source mmdet -m pytest tests/
> tests/test_models/test_forward.py 
> Error: Process completed with exit code 137.
> ```
> 
> I do not quite understand what is the reason for it and how to fix it.

Hi, we have fixed the CI in #5862. Could you merge the master branch?

**Comment by ZwwWayne** [comment]
Since [MMDeploy](https://github.com/open-mmlab/mmdeploy) already supports OpenVINO. This PR will no longer be continued and is thus closed.

**Pull request closed without merging**